pkg/utils/selector.go

```diff
@@ -0,0 +1,118 @@
+package utils
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+type manifestPath []string
+
+var (
+	// known manifest/resource paths for selector requirements
+	selectorMatchLabels         manifestPath = []string{"spec", "selector", "matchLabels"}
+	podSelectorMatchLabels      manifestPath = []string{"spec", "podSelector", "matchLabels"}
+	selectorMatchExpressions    manifestPath = []string{"spec", "selector", "matchExpressions"}
+	podSelectorMatchExpressions manifestPath = []string{"spec", "podSelector", "matchExpressions"}
+)
+
+// GetLabelSelector parses the unstructured object and checks following:
+// 1. Is there some "spec.selector" or "spec.podSelector". If not then return nil.
+// 2. Is there some empty "spec.selector" or "spec.podSelector". If yes then return empty LabelSelector struct.
+// 3. Otherwise try to parse "matchLabels" and "matchExpressions" requirements and create LabelSelector with
+// these values.
+func GetLabelSelector(object *unstructured.Unstructured) *metav1.LabelSelector {
+	if !hasSelector(object) {
+		return nil
+	}
+
+	if hasEmptySelector(object) {
+		return &metav1.LabelSelector{}
+	}
+
+	matchLabels := readMatchLabels(object)
+	matchExpressions := readMatchExpressions(object)
+	return &metav1.LabelSelector{
+		MatchExpressions: matchExpressions,
+		MatchLabels:      matchLabels,
+	}
+}
+
+// readMatchLabels tries to parse known paths for the "matchLabels" attribute in the
+// unstructured object input
+func readMatchLabels(object *unstructured.Unstructured) map[string]string {
+	for _, path := range []manifestPath{selectorMatchLabels, podSelectorMatchLabels} {
+		pathExists := pathExistsAsMap(object, path)
+		if pathExists {
+			matchLabels, _, _ := unstructured.NestedStringMap(object.Object, path...)
+			return matchLabels
+		}
+	}
+	return nil
+}
+
+// readMatchExpressions tries to parse known paths for the "matchExpressions" attribute in the
+// unstructured object input
+func readMatchExpressions(object *unstructured.Unstructured) []metav1.LabelSelectorRequirement {
+	matchExpressionsRequirements := []metav1.LabelSelectorRequirement{}
+	for _, path := range []manifestPath{selectorMatchExpressions, podSelectorMatchExpressions} {
+		pathExists := pathExistsAsSlice(object, path)
+		if pathExists {
```

Comment:
nit: this function could be less complex just inverting the conditional.
Instead of:
```
if true {
    // stuff with depth
}
```
You can just write:
```
if !pathExists {
    continue
}
// stuff here
```

It's not big deal but with a function like this (that contains 4 depth tabs) it could make the code easier to read for a cheap fast exit.

Reply:
Yeah sure. OK. Thanks